### DISCUSSION TITLE
An easy way to copy records with relations?

### DISCUSSION THREAD

**Discussion by ilja-denisovs** [prisma/prisma#18772] [Q&A] [answered]
Hi! 

Is there an easy way to copy records with relations? Especially those ones, which have relations with Json kind fields. 

See my example Schema:

```
model User {
  id       Int        @id @default(autoincrement())
  name     String?
  Contacts Contacts[]
  Devices  Devices[]
}

model Devices {
  id     Int    @id @default(autoincrement())
  brand  String
  name   Json
  price  Float  @default(0.0)
  userId Int
  user   User   @relation(fields: [userId], references: [id])
}
```

When I'm trying to copy it like this:

```typescript
type UserCopyInput = User & {
  Devices: Devices[]
};

async function copyUser(user: UserCopyInput) {
  const userCopy = {
    ...user,
    Devices: {
      create: user.Devices
    }
  };

  await prisma.user.create({
    data: userCopy
  });
}
```

It throws  `Type 'JsonValue' is not assignable to type 'JsonNull | InputJsonValue'.` error.

When I trying to redefine the Json field, like in following example: 

```typescript
async function copyUser(user: UserCopyInput) {
  const devicesCopy = user.Devices.map((device) => ({
    ...device,
    name: device.name as object
  }));

  const userCopy = {
    ...user,
    Devices: {
      create: devicesCopy
    }
  };

  await prisma.user.create({
    data: userCopy
  });
}
```

But it throws the `Unknown arg `userId` in data.Devices.create.0.userId for type DevicesUncheckedCreateWithoutUserInput.` error. 

Finally, I found a solution which works, but as in real life project I had a way longer tables, I supposed to produce a tons of code in order to make such a simple action, so I just wondering, if there is an couple-of-lines way to copy existing records with relations?

Thank you!

**Comment by ilja-denisovs** [comment]
Solution example, that worked for me: 

```typescript
async function copyUser(user: UserCopyInput) {
  const devicesCopy = user.Devices.map((device) => {
    const { id, userId, ...deviceCopy } = device;

    return {
      ...deviceCopy,
      name: deviceCopy.name as object
    };
  });

  const { id, ...userCopy } = user;

  await prisma.user.create({
    data: {
      ...userCopy,
      Devices: {
        create: devicesCopy
      }
    }
  });
}
```

**Comment by ludralph** [comment, marked as answer]
Hi @ilja-denisovs 👋 

There is no built-in method to create a copy of a record with relations. You would need to write custom code to fit your use case like you're already doing. Also, one way to have handled the JSON error was to use both the `JSON.parse` method and `JSON.stringify` method as shown below
```js
async function copyUser(user: UserCopyInput) {
  const devicesCopy = user.Devices.map((device) => ({
    ...device,
    name: JSON.parse(JSON.stringify(device.name))
  }));

  const userCopy = {
    ...user,
    Devices: {
      create: devicesCopy
    }
  };

  await prisma.user.create({
    data: userCopy
  });
}
```

**Reply by ilja-denisovs** [reply]
Hi @ludralph! Thank you for fast reply. I'll keep in mind this solution.

**Reply by ludralph** [reply]
Hi @ilja-denisovs 

Do you have any more questions around this issue? If no, it would be really helpful if you could quickly **mark this Discussion as _answered_** to indicate that your issue has been resolved. 

In case you do have any outstanding questions, feel free to let us know and we'd be happy to help!
